server/src/main/java/io/crate/replication/logical/MetadataTracker.java

```diff
@@ -284,6 +294,129 @@ static ClusterState updateIndexMetadata(String subscriptionName,
+                                                ClusterState publisherClusterState,
+                                                LogicalReplicationService replicationService) {
+        var subscription = replicationService.subscriptions().get(subscriptionName);
+        assert subscription != null : "Cannot retrieve subscription '" + subscriptionName + "' from local cluster state";
+        subscribeToNewRelations(
+            subscriptionName,
+            subscriberClusterState,
+            publisherClusterState,
+            relationNames -> replicationService.updateSubscriptionState(
+                subscriptionName,
+                relationNames,
+                Subscription.State.INITIALIZING,
+                null
+            ),
+            (relationNames, toRestoreIndices, toRestoreTemplates) -> replicationService.restore(
+                subscriptionName,
+                subscription.settings(),
+                relationNames,
+                toRestoreIndices,
+                toRestoreTemplates
+            )
+        );
+    }
+
+    @VisibleForTesting
+    static CompletableFuture<Boolean> subscribeToNewRelations(
+        String subscriptionName,
+        ClusterState subscriberClusterState,
+        ClusterState publisherClusterState,
+        Function<Collection<RelationName>, CompletableFuture<Boolean>> updateStateFunction,
+        TriConsumer<Collection<RelationName>, List<String>, List<String>> restoreConsumer) {
+        PublicationsMetadata publicationsMetadata = publisherClusterState.metadata().custom(PublicationsMetadata.TYPE);
+        SubscriptionsMetadata subscriptionsMetadata = subscriberClusterState.metadata().custom(SubscriptionsMetadata.TYPE);
+        if (publicationsMetadata == null || subscriptionsMetadata == null) {
+            return CompletableFuture.completedFuture(true);
+        }
+        Subscription subscription = subscriptionsMetadata.subscription().get(subscriptionName);
+
+        var subscribedRelations = subscription.relations();
+        var publications = publicationsMetadata.publications();
+        var relationNamesForStateUpdate = new HashSet<RelationName>();
+        var toRestoreIndices = new ArrayList<String>();
+        var toRestoreTemplates = new ArrayList<String>();
+        for (var publicationName : subscription.publications()) {
+            var publication = publications.get(publicationName);
+
+            Predicate<String> skipIndexOrTemplate = indexOrTemplateName ->
+                subscriberClusterState.metadata().hasIndex(indexOrTemplateName)
+                || subscriberClusterState.metadata().templates().containsKey(indexOrTemplateName);
+            if (publication.isForAllTables() == false) {
+                skipIndexOrTemplate = skipIndexOrTemplate.and(
+                    indexOrTemplateName -> publication.tables().contains(RelationName.fromIndexName(indexOrTemplateName))
```

Comment:
Maybe change the predicate to take a `relationName` directly to avoid creating the `IndexParts` twice in the true case (see loop further below - it creates IndexParts/relationName - which is the same as `RelationName.fromIndexName` does.

Reply:
The templateName cannot be resolved by a `RelationName`, but I've changed the predicate to `BiPredicate<String, RelationName>` to avoid creating the `relationName` twice. Thanks for pointing to that, had same in mind, but didn't think of a `BiPredicate` in the first place.